### PR TITLE
Run kubelet outside container

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What this PR does / why we need it**: In current (v0.14.6) acs-engine implementation, kubelet starts in a Docker container. 
There are a lot of regressions of containerized kubelet reported in Kubernetes v1.10 (https://github.com/kubernetes/kubernetes/issues/61456) . Since there are no full e2e tests for this, it may also happen in future releases (thanks @feiskyer for reporting). This PR moves kubelet to run outside the container, which is expected to be more stable.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**If applicable**:
- [ ] documentation
- [ ] unit tests
- [x] tested backward compatibility (ie. deploy with previous version, upgrade with this branch)

**Release note**:

```release-note
```

### CODE REVIEW

File: parts/k8s/kubernetesmastercustomdata.yml [hunk 1]
```diff
@@ -137,11 +137,13 @@ MASTER_ADDONS_CONFIG_PLACEHOLDER
     ExecStartPre=/usr/bin/docker pull {{WrapAsVariable "kubernetesHyperkubeSpec"}}
     ExecStartPre=/usr/bin/docker run --rm -v /tmp/kubectldir:/opt/kubectldir {{WrapAsVariable "kubernetesHyperkubeSpec"}} /bin/bash -c "cp /hyperkube /opt/kubectldir/"
 {{if .MasterProfile.IsCoreOS}}
+    ExecStartPre=/bin/cp /tmp/kubectldir/hyperkube /opt/kubelet
     ExecStartPre=/bin/mv /tmp/kubectldir/hyperkube /opt/kubectl
-    ExecStart=/bin/chmod a+x /opt/kubectl
+    ExecStart=/bin/chmod a+x /opt/kubelet && /bin/chmod a+x /opt/kubectl
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetesmastercustomdata.yml]: TODO: adapt kubelet start for coreos filepath

File: parts/k8s/kubernetesagentcustomdata.yml [hunk 2]
```diff
@@ -96,11 +96,13 @@ write_files:
...
 {{if .IsCoreOS}}
+    ExecStartPre=/bin/cp /tmp/kubectldir/hyperkube /opt/kubelet
     ExecStartPre=/bin/mv /tmp/kubectldir/hyperkube /opt/kubectl
-    ExecStart=/bin/chmod a+x /opt/kubectl
+    ExecStart=/bin/chmod a+x /opt/kubelet && /bin/chmod a+x /opt/kubectl
 {{else}}
+    ExecStartPre=/bin/cp /tmp/kubectldir/hyperkube /usr/local/bin/kubelet
     ExecStartPre=/bin/mv /tmp/kubectldir/hyperkube /usr/local/bin/kubectl
-    ExecStart=/bin/chmod a+x /usr/local/bin/kubectl
+    ExecStart=/bin/chmod a+x /usr/local/bin/kubelet && /bin/chmod a+x /usr/local/bin/kubectl
```
Comment (jackfrancis) [inline parts/k8s/kubernetesagentcustomdata.yml]: We can shorten this to `/bin/chmod a+x /usr/local/bin/kubelet /usr/local/bin/kubectl` (the build of `chmod` on ubuntu hosts accepts a list of file targets to apply the mode changes against). Verified:

```
azureuser@k8s-master-18440544-0:/tmp$ ls -la test*
-rw-rw-r-- 1 azureuser azureuser 0 Apr  3 23:08 test1
-rw-rw-r-- 1 azureuser azureuser 0 Apr  3 23:08 test2
azureuser@k8s-master-18440544-0:/tmp$ /bin/chmod a+x test1 test2
azureuser@k8s-master-18440544-0:/tmp$ ls -la test*
-rwxrwxr-x 1 azureuser azureuser 0 Apr  3 23:08 test1
-rwxrwxr-x 1 azureuser azureuser 0 Apr  3 23:08 test2
```

File: parts/k8s/kubernetesagentcustomdata.yml [hunk 3]
```diff
@@ -96,11 +96,13 @@ write_files:
     ExecStartPre=/usr/bin/docker pull {{WrapAsVariable "kubernetesHyperkubeSpec"}}
     ExecStartPre=/usr/bin/docker run --rm -v /tmp/kubectldir:/opt/kubectldir {{WrapAsVariable "kubernetesHyperkubeSpec"}} /bin/bash -c "cp /hyperkube /opt/kubectldir/"
 {{if .IsCoreOS}}
+    ExecStartPre=/bin/cp /tmp/kubectldir/hyperkube /opt/kubelet
     ExecStartPre=/bin/mv /tmp/kubectldir/hyperkube /opt/kubectl
-    ExecStart=/bin/chmod a+x /opt/kubectl
+    ExecStart=/bin/chmod a+x /opt/kubelet && /bin/chmod a+x /opt/kubectl
```
Comment (jackfrancis) [inline parts/k8s/kubernetesagentcustomdata.yml]: See below. Didn't verify on a coreos host, but I'm confident this is a standard pattern.

File: parts/k8s/kubernetesmastercustomdata.yml (hunk 1 above)
Comment (jackfrancis) [inline parts/k8s/kubernetesmastercustomdata.yml]: ditto above about chmod command

File: parts/k8s/kubernetesmastercustomdata.yml [hunk 4]
```diff
@@ -137,11 +137,13 @@ MASTER_ADDONS_CONFIG_PLACEHOLDER
...
 {{else}}
+    ExecStartPre=/bin/cp /tmp/kubectldir/hyperkube /usr/local/bin/kubelet
     ExecStartPre=/bin/mv /tmp/kubectldir/hyperkube /usr/local/bin/kubectl
-    ExecStart=/bin/chmod a+x /usr/local/bin/kubectl
+    ExecStart=/bin/chmod a+x /usr/local/bin/kubelet && /bin/chmod a+x /usr/local/bin/kubectl
```
Comment (jackfrancis) [inline parts/k8s/kubernetesmastercustomdata.yml]: ditto

File: parts/k8s/kubernetesmastercustomdata.yml [hunk 5]
```diff
@@ -137,11 +137,13 @@ MASTER_ADDONS_CONFIG_PLACEHOLDER
     ExecStartPre=/usr/bin/docker pull {{WrapAsVariable "kubernetesHyperkubeSpec"}}
     ExecStartPre=/usr/bin/docker run --rm -v /tmp/kubectldir:/opt/kubectldir {{WrapAsVariable "kubernetesHyperkubeSpec"}} /bin/bash -c "cp /hyperkube /opt/kubectldir/"
 {{if .MasterProfile.IsCoreOS}}
+    ExecStartPre=/bin/cp /tmp/kubectldir/hyperkube /opt/kubelet
     ExecStartPre=/bin/mv /tmp/kubectldir/hyperkube /opt/kubectl
-    ExecStart=/bin/chmod a+x /opt/kubectl
+    ExecStart=/bin/chmod a+x /opt/kubelet && /bin/chmod a+x /opt/kubectl
 {{else}}
+    ExecStartPre=/bin/cp /tmp/kubectldir/hyperkube /usr/local/bin/kubelet
```
Comment (jackfrancis) [inline parts/k8s/kubernetesmastercustomdata.yml]: Let's do this in one line to make it a little clearer that we are copying the same file to two different places. One example:

`for dest in /usr/local/bin/kubelet /usr/local/bin/kubectl ; do cp /tmp/kubectldir/hyperkube $dest; done`

And then just add a cleanup line after, like:

`rm -Rf /tmp/kubectldir`

(but replace kubectldir with the new, general name in both instances :) )

Comment (CecileRobertMichon) [reply parts/k8s/kubernetesmastercustomdata.yml]: I don't really find that
 ```
ExecStartPre=for dest in /usr/local/bin/kubelet /usr/local/bin/kubectl ; do /bin/cp /tmp/hyperkube/hyperkube $dest; done
ExecStartPre=rm -Rf /tmp/kubectldir
```

is much more readable than
```
ExecStartPre=/bin/cp /tmp/hyperkubedir/hyperkube /usr/local/bin/kubelet
ExecStartPre=/bin/mv /tmp/hyperkubedir/hyperkube /usr/local/bin/kubectl
```

How about: 
```
ExecStartPre=/bin/mv /tmp/hyperkubedir/hyperkube /usr/local/bin/kubelet
ExecStartPre=/bin/cp /usr/local/bin/kubelet /usr/local/bin/kubectl
```
?

Comment (CecileRobertMichon) [reply parts/k8s/kubernetesmastercustomdata.yml]: I mean I see where you're coming from in terms of logic, I just find that it makes the syntax more complex to have a for loop for only two files. Please feel free to disagree with me :)

File: parts/k8s/kubernetesagentcustomdata.yml (hunk 2 above)
Comment (CecileRobertMichon) [reply parts/k8s/kubernetesagentcustomdata.yml]: Yes good catch, thanks

File: parts/k8s/kubernetesagentcustomdata.yml [hunk 6]
```diff
@@ -92,15 +92,17 @@ write_files:
     TimeoutStartSec=0
     Restart=on-failure
     RestartSec=5s
-    ExecStartPre=/bin/mkdir -p /tmp/kubectldir
+    ExecStartPre=/bin/mkdir -p /tmp/hyperkubedir
     ExecStartPre=/usr/bin/docker pull {{WrapAsVariable "kubernetesHyperkubeSpec"}}
-    ExecStartPre=/usr/bin/docker run --rm -v /tmp/kubectldir:/opt/kubectldir {{WrapAsVariable "kubernetesHyperkubeSpec"}} /bin/bash -c "cp /hyperkube /opt/kubectldir/"
+    ExecStartPre=/usr/bin/docker run --rm -v /tmp/hyperkubedir:/opt/hyperkubedir {{WrapAsVariable "kubernetesHyperkubeSpec"}} /bin/bash -c "cp /hyperkube /opt/hyperkubedir/"
```
Comment (feiskyer) [inline parts/k8s/kubernetesagentcustomdata.yml]: Besides the extraction, dependencies should also be installed. e.g. on ubuntu, kubelet depends on `iptables (>= 1.4.21), kubernetes-cni (= 0.6.0), iproute2, socat, util-linux, mount, ebtables, ethtool, init-system-helpers (>= 1.18~)`

Comment (CecileRobertMichon) [reply parts/k8s/kubernetesagentcustomdata.yml]: Thanks @feiskyer. Just checked and all of these are already being installed except for `socat`. I will add a step to install them if they are missing so we are safe. I can't find a package `kubernetes-cni` however, do you have more info on that one?

File: parts/k8s/artifacts/1.5/kuberneteskubelet.service [hunk 7]
```diff
@@ -1,45 +0,0 @@
-[Unit]
```
Comment (CecileRobertMichon) [inline parts/k8s/artifacts/1.5/kuberneteskubelet.service]: Removed the 1.5 version since k8s 1.5 was deprecated in #2394 and it's not in use anymore

File: parts/k8s/kubernetesagentcustomdata.yml [hunk 8]
```diff
@@ -128,7 +130,9 @@ AGENT_ARTIFACTS_CONFIG_PLACEHOLDER
   owner: "root"
   content: |
     #!/bin/bash
-
+    . /opt/azure/containers/provision_source.sh
+    apt_get_update
+    retrycmd_if_failure 20 10 120 apt-get install -y iptables iproute2 socat util-linux mount ebtables ethtool init-system-helpers
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetesagentcustomdata.yml]: Ensure dependencies are installed on agents

File: parts/k8s/kubernetesmastercustomdata.yml [hunk 9]
```diff
@@ -177,6 +179,9 @@ MASTER_ARTIFACTS_CONFIG_PLACEHOLDER
   content: |
     #!/bin/bash
     set -e
+    . /opt/azure/containers/provision_source.sh
+    apt_get_update
+    retrycmd_if_failure 20 10 120 apt-get install -y iptables iproute2 socat util-linux mount ebtables ethtool init-system-helpers
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetesmastercustomdata.yml]: Ensure dependencies are installed on masters

File: parts/k8s/kubernetesagentcustomdata.yml (hunk 6 above)
Comment (CecileRobertMichon) [reply parts/k8s/kubernetesagentcustomdata.yml]: For the cni plugin, I think we are already installing it at https://github.com/Azure/acs-engine/blob/b463cd0b0e5818e9da87ef75a86dc45ac14dc014/parts/k8s/kubernetesmastercustomscript.sh#L217

File: parts/k8s/kubernetesagentcustomdata.yml [hunk 10]
```diff
@@ -128,7 +129,9 @@ AGENT_ARTIFACTS_CONFIG_PLACEHOLDER
   owner: "root"
   content: |
     #!/bin/bash
-
+    source /opt/azure/containers/provision_source.sh
+    apt_get_update
+    retrycmd_if_failure 20 10 120 apt-get install -y iptables iproute2 socat util-linux mount ebtables ethtool init-system-helpers
```
Comment (jackfrancis) [inline parts/k8s/kubernetesagentcustomdata.yml]: Are we confident that all of these packages will reliably install in under 2 mins? (speaking to the 120 second timeout)

Comment (CecileRobertMichon) [reply parts/k8s/kubernetesagentcustomdata.yml]: Yes most of them should already be there so it's very quick. In my tests it took under 10 seconds. I can increase the timeout if we want to be sure.

File: parts/k8s/kubernetesprovisionsource.sh [hunk 11]
```diff
@@ -2,4 +2,6 @@
 
 retrycmd_if_failure() { retries=$1; wait=$2; timeout=$3; shift && shift && shift; for i in $(seq 1 $retries); do timeout $timeout ${@}; [ $? -eq 0  ] && break || sleep $wait; done; echo Executed \"$@\" $i times; }
 retrycmd_if_failure_no_stats() { retries=$1; wait=$2; timeout=$3; shift && shift && shift; for i in $(seq 1 $retries); do timeout $timeout ${@}; [ $? -eq 0  ] && break || sleep $wait; done; }
-retrycmd_get_tarball() { retries=$1; wait=$2; tarball=$3; url=$4; for i in $(seq 1 $retries); do tar -tzf $tarball; [ $? -eq 0  ] && break || retrycmd_if_failure_no_stats $retries 1 10 curl -fsSL $url -o $tarball; sleep $wait; done; }
\ No newline at end of file
+retrycmd_get_tarball() { retries=$1; wait=$2; tarball=$3; url=$4; for i in $(seq 1 $retries); do tar -tzf $tarball; [ $? -eq 0  ] && break || retrycmd_if_failure_no_stats $retries 1 10 curl -fsSL $url -o $tarball; sleep $wait; done; }
+ensure_etcd_ready() { for i in $(seq 1 1800); do if [ -e /opt/azure/containers/certs.ready ]; then break; fi; sleep 1; done }
```
[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply parts/k8s/kubernetesprovisionsource.sh]: Oh yeah I just moved it out because I moved out apt_get_update so I thought might as well but we should generalize it in another PR

[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply parts/k8s/kubernetesprovisionsource.sh]: I increased it to 5 minutes... Should be plenty enough
